File: examples/c/mcc152/analog_output_write/analog_output_write.c
/*****************************************************************************

    MCC 152 Functions Demonstrated:
        mcc152_a_out_write
        mcc152_info

    Purpose:
        Write values to analog output 0 in a loop.

    Description:
        This example demonstrates writing output data using analog output 0.

*****************************************************************************/
#include <stdbool.h>
#include "../../daqhats_utils.h"

#define CHANNEL     0               // output channel, set to 1 for channel 1
#define OPTIONS     OPTS_DEFAULT    // default output options (voltage), set to 
                                    // OPTS_NOSCALEDATA to use DAC codes

#define BUFFER_SIZE 32

bool get_input_value(double* p_value)
{
    char buffer[BUFFER_SIZE];
    double min;
    double max;
    double value;

    if (p_value == NULL)
    {
        return false;
    }
    
    // Get the min and max voltage/code values for the analog outputs to 
    // validate the user input.
    if ((OPTIONS & OPTS_NOSCALEDATA) == 0)
    {
        min = mcc152_info()->AO_MIN_RANGE;
        max = mcc152_info()->AO_MAX_RANGE;
    }
    else
    {
        min = (double)mcc152_info()->AO_MIN_CODE;
        max = (double)mcc152_info()->AO_MAX_CODE;
    }

    while (1)
    {
        if ((OPTIONS & OPTS_NOSCALEDATA) == 0)
        {
            printf("Enter a value between %.1f and %.1f, non-numeric character "
                "to exit: ", min, max);
        }
        else
        {
            printf("Enter a value between %.0f and %.0f, non-numeric character "
                "to exit: ", min, max);
        }

        if (fgets(buffer, BUFFER_SIZE, stdin) == NULL)
        {
            // Empty string or input error.
            return false;
        }
        else
        {
            if (buffer[0] == '\n')
            {
                // Enter.
                return false;
            }
            
            // Got a string - convert to a number.
            if (sscanf(buffer, "%lf", &value) == 0)
            {
                // Not a number.
                return false;
            }
            
            // Compare the number to min and max allowed.
            if ((value < min) || (value > max))
            {
                // Out of range, ask again.
                printf("Value out of range.\n");
            }
            else
            {
                // Valid value.
                *p_value = value;
                return true;
            }
        }
    }
}

int main(void)
{
    uint8_t address;
    int result = RESULT_SUCCESS;
    double value;
    bool run_loop;
    char options_str[256];

    printf("\nMCC 152 single channel analog output example.\n");
    printf("Writes the entered value to the analog output.\n");
    printf("   Functions demonstrated:\n");
    printf("      mcc152_a_out_write\n");
    printf("      mcc152_info\n");
    printf("   Channel: %d\n", CHANNEL);

    convert_options_to_string(OPTIONS, options_str);
    printf("   Options: %s\n\n", options_str);

    // Select the device to be used.
    if (select_hat_device(HAT_ID_MCC_152, &address) != 0)
    {
        return 1;
    }
    
    printf("\nUsing address %d.\n", address);
    
    // Open a connection to the device.
    result = mcc152_open(address);
    if (result != RESULT_SUCCESS)
    {
        print_error(result);
        // Could not open the device - exit.
        printf("Unable to open device at address %d\n", address);
        return 1;
    }
        
    run_loop = true;
    // Loop until the user terminates or we get a library error.
    while (run_loop)
    {
        // Get the value from the user.
        if (get_input_value(&value))
        {
            // Write the value to the selected channel.
            result = mcc152_a_out_write(address, CHANNEL, OPTIONS, value);
            if (result != RESULT_SUCCESS)
            {
                print_error(result);
                mcc152_close(address);
                return 1;
            }
        }
        else
        {
            run_loop = false;
        }
    }

    // Reset the output to 0V.
    result = mcc152_a_out_write(address, CHANNEL, OPTIONS, 0.0);
    if (result != RESULT_SUCCESS)
    {
        print_error(result);
        mcc152_close(address);
        return 1;
    }
    
    // Close the device.
    result = mcc152_close(address);
    if (result != RESULT_SUCCESS)
    {
        print_error(result);
        return 1;
    }

    return 0;
}

File: examples/python/mcc152/analog_output_write.py
#!/usr/bin/env python
"""
    MCC 152 Methods Demonstrated:
        mcc152.a_out_write
        mcc152.info

    Purpose:
        Write a single data value for channel 0 in a loop.

    Description:
        This example demonstrates writing output data using analog output 0.
"""
from __future__ import print_function
from sys import version_info
from daqhats import mcc152, OptionFlags, HatIDs, HatError
from daqhats_utils import select_hat_device

def get_input_value():
    """
    Get the voltage from the user and validate it.
    """

    # Get the min and max voltage values for the analog outputs to validate
    # the user input.
    min_v = mcc152.info().AO_MIN_RANGE
    max_v = mcc152.info().AO_MAX_RANGE

    while True:
        message = ("Enter a voltage between {0:.1f} and {1:.1f}, "
                   "non-numeric character to exit: ".format(min_v, max_v))

        if version_info.major > 2:
            str_v = input(message)
        else:
            str_v = raw_input(message)

        try:
            value = float(str_v)
        except ValueError:
            raise
        else:
            if (value < min_v) or (value > max_v):
                # Out of range, ask again.
                print("Value out of range.")
            else:
                # Valid value.
                return value

def main():
    """
    This function is executed automatically when the module is run directly.
    """
    options = OptionFlags.DEFAULT
    channel = 0
    num_channels = mcc152.info().NUM_AO_CHANNELS

    # Ensure channel is valid.
    if channel not in range(num_channels):
        error_message = ('Error: Invalid channel selection - must be '
                         '0 - {}'.format(num_channels - 1))
        raise Exception(error_message)

    print('MCC 152 single channel analog output example.')
    print('Writes the specified voltage to the analog output.')
    print("   Methods demonstrated:")
    print("      mcc152.a_out_write")
    print("      mcc152.info")
    print("   Channel: {}\n".format(channel))

    # Get an instance of the selected hat device object.
    address = select_hat_device(HatIDs.MCC_152)

    print("\nUsing address {}.\n".format(address))

    hat = mcc152(address)

    run_loop = True
    error = False
    while run_loop and not error:
        # Get the value from the user.
        try:
            value = get_input_value()
        except ValueError:
            run_loop = False
        else:
            # Write the value to the selected channel.
            try:
                hat.a_out_write(channel=channel,
                                value=value,
                                options=options)
            except (HatError, ValueError):
                error = True

    if error:
        print("Error writing analog output.")

if __name__ == '__main__':
    # This will only be run when the module is called directly.
    main()
